### PR TITLE
Save media file considering the file extension

### PR DESCRIPTION
**Pull request by sangwoo108**
We were saving media files assuming that they're all .mp4 files. But some contents give us .m3u8 files so we should save files considering the original file extension

Resolves https://github.com/brave/brave-browser/issues/30102

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
PlaylistServiceUnitTest.MediaFileextension

### CODE REVIEW

File: components/playlist/browser/playlist_service.cc [hunk 1]
```diff
@@ -989,6 +994,13 @@ bool PlaylistService::IsValidPlaylistItem(const std::string& id) {
   return HasPrefStorePlaylistItem(id);
 }
 
+base::FilePath PlaylistService::GetMediaPathForPlaylistItemItem(
+    const std::string& id) {
+  base::FilePath path;
+  CHECK(GetMediaPath(id, &path));
```
Comment (simonhong) [inline components/playlist/browser/playlist_service.cc:1000]: Using `CHECK()` here ok? It seems `GetMediaPath()` could give `false`.

File: components/playlist/browser/playlist_service.cc [hunk 2]
```diff
@@ -1049,10 +1061,34 @@ bool PlaylistService::GetThumbnailPath(const std::string& id,
...
+  if (HasPlaylistItem(id)) {
+    // Try to infer file extension from the source URL.
+    auto item = GetPlaylistItem(id);
+    GURL url(item->media_source);
+    auto path = url.path();
+    if (!path.empty()) {
+      if (path.at(0) == '/') {
+        // Drop the leading '/'.
+        path = std::string(path.begin() + 1, path.end());
+      }
```
Comment (simonhong) [inline components/playlist/browser/playlist_service.cc:1078]: I think we don't need this because below code only splits with `.` and only takes last part?

Comment (sangwoo108) [reply components/playlist/browser/playlist_service.cc]: Yeah, make sense. Removed.

File: components/playlist/browser/playlist_service.cc (hunk 1 above)
Comment (sangwoo108) [reply components/playlist/browser/playlist_service.cc:1000]: Yeah, that's what I'm not sure these days. Upstream folks now recommend CHECK over DCHECK(https://bravesoftware.slack.com/archives/CU5UXM3EC/p1683249977446639). And this method should be called for a valid item, assertion seems to be better than early return.

**Review by simonhong — APPROVED** [review]
++

**Comment by sangwoo108** [comment]
@deeppandya Could you confirm that this doesn't break Android behaviors?

**Comment by sangwoo108** [comment]
@deeppandya confirmed that this wasn't enough for us. I'm trying peeking mime type from download response.

**Comment by sangwoo108** [comment]
@simonhong , Sorry to bother you but I updated PR to get more precise extension from response header. Reportedly, some sites don't have proper file extension at the end of source url, so we need this mechanism as a fallback. Would you mind taking another look?

**Review by simonhong — COMMENTED** [review]
++ still looks good 👍🏼

**Review by deeppandya — APPROVED** [review]
LGTM. Verified the PR changes with android changes. Works as expected for downloaded file.

**Pull request merged**